### PR TITLE
EZP-25009: Role assignement with section & subtree limitations

### PR DESCRIPTION
**Pull request by StephaneDiot**
jira: https://jira.ez.no/browse/EZP-25009

## Description

The goal is to be able to assign a role to a user or user group with a section and subtree limitation.

:warning: the redirection is not implemented after the role is assigned. Seems it will wait for https://jira.ez.no/browse/EZP-25442

:warning:² <s>Currently it is working with IDs instead of URIs for REST request to assign role to user with limitation. It's not what we are expecting from the REST API. This will be changed when https://github.com/ezsystems/ezpublish-kernel/pull/1569 will be finnish. So it shouldn't be merged before.</s>  Now it's done !
## Tests

- Manual tests
- Unit tests

## TODO
- [x] replace IDs by URIs

### CODE REVIEW

**Comment by StephaneDiot** [comment]
ping @dpobel @yannickroger

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 1]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
         /**
          * tap event handler on the role assign buttons. It launches the
          * universal discovery widget so that the user can pick some contents.
+         * It also fill the config with datas concerning the section limitation
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: "fills with data"

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 2]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+         * @method _pickSubtreeWithSectionLimitation
+         * @protected
+         * @param {EventFacade} e
+         */
+        _pickSubtreeWithSectionLimitation: function (e) {
+            var button = e.target,
+                container = this.get('container'),
+                sectionSelector = container.one(".ez-role-assignment-section-id"),
+                sectionSelectedIndex = sectionSelector.get('selectedIndex'),
+                unsetLoading = Y.bind(this._uiUnsetUDWButtonLoading, this, button),
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: maybe also create a variable for: `sectionSelector.get('options').item(sectionSelectedIndex)`

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 3]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
          * @method _pickSubtree
          * @protected
          * @param {EventFacade} e
          */
         _pickSubtree: function (e) {
             var button = e.target,
-                unsetLoading = Y.bind(this._uiUnsetUDWButtonLoading, this, button);
+                unsetLoading = Y.bind(this._uiUnsetUDWButtonLoading, this, button),
+                udwConfigData = {
+                    roleId: button.getAttribute('data-role-rest-id'),
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: `'data-role-rest-id'` and 'data-role-name' could be stored in a variable so you could avoid duplicate the strings in the module.

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 4]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                    afterUpdateCallback: unsetLoading,
+                };
 
             e.preventDefault();
             this._uiSetUDWButtonLoading(button);
+            this._fireContentDiscover(button, unsetLoading, udwConfigData);
+        },
+
+        /**
+         * tap event handler for assigning role with subtree limitation. It launches the
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: Once a great nitpicker pointed out the lack of capital letter :)

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 5]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+
+            e.preventDefault();
+            this._uiSetUDWButtonLoading(button);
+            this.fire('contentDiscover', {
+                config: {
+                    title: button.getAttribute('data-universaldiscovery-limit-subtree-title'),
+                    cancelDiscoverHandler: unsetLoading,
+                    multiple: true,
+                    contentDiscoveredHandler: function(e) {
+                        that._setSubtreeLimitation(button, this, e);
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: Y.bind should do the trick.

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 6]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+         * The UDW config datas will be filled by the role infos and the locations of the subtree limitation chosen before.
+         *
+         * @method _pickSubtreeLimitation
+         * @protected
+         * @param {EventFacade} e
+         */
+        _setSubtreeLimitation: function (button, udView, e) {
+            var unsetLoading = Y.bind(this._uiUnsetUDWButtonLoading, this, button),
+                that = this,
+                selectedLocationsIds = Y.Array.map(e.selection, function(struct) {
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: Indentation is weird

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 7]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                    roleId: button.getAttribute('data-role-rest-id'),
+                    roleName: button.getAttribute('data-role-name'),
+                    afterUpdateCallback: unsetLoading,
+                    limitationType: 'Subtree',
+                    subtreeIds: selectedLocationsIds,
+                },
+                udwAfterActiveChangeEvent = udView.onceAfter('activeChange', function() {
+                    udwAfterActiveChangeEvent.detach();
+                        setTimeout(function() {
+                            that._fireContentDiscover(button, unsetLoading, udwConfigData);
```
Comment (yannickroger) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: 2 Y.bind ? ping @dpobel

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 8]
```diff
@@ -87,18 +93,89 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
+        /**
+         * Create the limitation object with the section currently chosen.
+         *
+         * @method _createSectionLimitation
+         * @protected
+         * @param {Object} data
+         * @return {Object} the limitation with its identifer and values as expected by the API
+         */
+        _createSectionLimitation: function (data) {
+            // limitation will be filled like example in UserService.prototype.newRoleAssignInputStruct()
```
Comment (yannickroger) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: Maybe something like: 
```javascript
// Creating a limitation in the format described in the JS REST client's `UserService.prototype.newRoleAssignInputStruct()`
```

Same below

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 9]
```diff
@@ -197,20 +270,25 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
          * @param {Function} callback the callback to call when other tasks are done
          */
-        _assignRoleCallback: function (roleId, roleName, contents, countAssigned, callback) {
+        _assignRoleCallback: function (assignActionData, contents, countAssigned, callback) {
             var notificationIdentifier = this._getAssignRoleNotificationIdentifier(
-                    'assign-role', roleId, contents
-                );
+                    'assign-role', assignActionData, contents
+                ),
+                notificationSucessText =  '"' + assignActionData.roleName + '" role has been assigned to ' + countAssigned + ' users/groups';
```
Comment (yannickroger) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: nitpick line too long

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 10]
```diff
@@ -37,27 +43,134 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                udwConfigData = {
+                    roleId: button.getAttribute('data-role-rest-id'),
+                    roleName: button.getAttribute('data-role-name'),
+                    afterUpdateCallback: unsetLoading,
+                    limitationType: 'Subtree',
+                    subtreeIds: selectedLocationsIds,
+                },
+                udwAfterActiveChangeEvent = udView.onceAfter('activeChange', function() {
+                    udwAfterActiveChangeEvent.detach();
+                        setTimeout(function() {
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: Indentation.

File: Tests/js/views/serverside/assets/ez-roleserversideview-tests.js [hunk 11]
```diff
@@ -189,6 +191,293 @@ YUI.add('ez-roleserversideview-tests', function (Y) {
...
+                method: 'get',
+                args: ['id'],
+                returns: subtreeLocationId
+            });
+
+            this.view.once('contentDiscover', function (e) {
+
+                that.resume(function() {
+                    e.config.contentDiscoveredHandler.call(udView, fakeEventFacade);
+                    //e.config.contentDiscoveredHandler.apply(udView, [selection]);
```
Comment (yannickroger) [inline Tests/js/views/serverside/assets/ez-roleserversideview-tests.js]: ooops

File: Resources/public/js/views/serverside/ez-roleserversideview.js (hunk 7 above)
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: That's an option. You can also set the context of the `onceAfter` function with its third parameter and indeed bind the `setTimeout` functon.

File: Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js [hunk 12]
```diff
@@ -168,32 +158,56 @@ YUI.add('ez-roleserversideviewservice-tests', function (Y) {
...
+            Mock.expect(this.userService, {
+                method: 'assignRoleToUserGroup',
+                args: [Mock.Value.String, this.roleAssignInputStruct, Mock.Value.Function],
+                run: function (userGroupURI, roleAssignInputStruct, cb) {
+                    Assert.areEqual(
+                        userGroupURI,
+                        '/api/ezp/v2/user/groups' + userGroupId,
+                        "The userGroupId should be defined"
+                    );
+                    cb(false);
```
Comment (yannickroger) [inline Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js]: What happens if true (rest call fails) ?

**Comment by yannickroger** [comment]
+1 with feedback taken into account

File: Resources/public/js/views/serverside/ez-roleserversideview.js (hunk 5 above)
Comment (StephaneDiot) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: I need to keep the udwView here (-> this), so I can't use Y.bind or I will loose the context of the udwView for the parameters of . _setSubtreeLimitation

**Comment by StephaneDiot** [comment]
@dpobel

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 13]
```diff
@@ -37,33 +45,140 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                    roleId: button.getAttribute(ROLE_REST_ID),
+                    roleName: button.getAttribute(ROLE_NAME),
+                    afterUpdateCallback: unsetLoading,
+                    limitationType: 'Subtree',
+                    subtreeIds: selectedLocationsIds,
+                },
+                udwAfterActiveChangeEvent = udView.onceAfter('activeChange', function() {
+                    udwAfterActiveChangeEvent.detach();
+                    setTimeout(Y.bind(function() {
+                        console.log(this);
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: ...

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 14]
```diff
@@ -37,33 +45,140 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                        this._fireContentDiscover(button, unsetLoading, udwConfigData);
+                        }, this), 0);
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: wrong indentation level

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 15]
```diff
@@ -87,18 +93,89 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
+         * @protected
+         * @param {Object} data
+         * @return {Object} the limitation with its identifer and values as expected by the API
+         */
+        _createSectionLimitation: function (data) {
+            // Creating a limitation in the format described in the JS REST client's `UserService.prototype.newRoleAssignInputStruct()`
+            var limitation = {
+                    "_identifier": data.limitationType,
+                    "values": {
+                        "ref":[{
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: nitpick: space after `:`

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 16]
```diff
@@ -87,18 +93,89 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
+            // Creating a limitation in the format described in the JS REST client's `UserService.prototype.newRoleAssignInputStruct()`
+            var limitation,
+                limitationValues = Y.Array.map(data.subtreeIds, function(subtreeId) {
+                    return {"_href": subtreeId, "_media-type": "application\/vnd.ez.api.Location+json"};
+                });
+
+            limitation = {
+                "_identifier": "Subtree",
+                "values": {
+                    "ref":limitationValues
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: nitpick: space afer `:`

File: Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js [hunk 17]
```diff
@@ -206,23 +220,141 @@ YUI.add('ez-roleserversideviewservice-tests', function (Y) {
...
+
+        "Should assign role with a subtree limitation to the user and call the callback": function () {
+            var universalDiscovery = new Mock(),
+                callbackCalled = false,
+                callback = function () {
+                    callbackCalled = true;
+                },
+                limitationType = 'Subtree',
+                subtreeIds = ['/1/2/'],
+                that = this;
```
Comment (dpobel) [inline Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js]: useless

File: Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js [hunk 18]
```diff
@@ -206,23 +220,141 @@ YUI.add('ez-roleserversideviewservice-tests', function (Y) {
...
+            Assert.isTrue(callbackCalled, 'The callback should be called');
+        },
+
+        "Should assign role to the user group and call the callback": function () {
+            var universalDiscovery = new Mock(),
+                callbackCalled = false,
+                callback = function () {
+                    callbackCalled = true;
+                },
+                that = this;
```
Comment (dpobel) [inline Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js]: useless as well

**Comment by dpobel** [comment]
+1 after fixing last inline comments

**Comment by andrerom** [comment]
🎉 😊 *(but I assume failure is irrelevant 😉)*

**Pull request merged**